inference-engine/src/gna_plugin/frontend/model_quantizer.hpp

```diff
@@ -100,14 +100,66 @@ class ModelQuantizer {
-            for (auto &&layer : sf.getStartLayers()) {
+        int infiniteLoopCount = 0;
+        std::vector<std::string> infiniteLoopPattern;
+        std::vector<std::string> infiniteLoopHistory;
+        while (!sf.allLayersProcessed() && infiniteLoopCount <= 2) {
+            auto layers = sf.getStartLayers();
+            infiniteLoopHistory.emplace_back(layers.front()->name);
+            for (auto &&layer : layers) {
                 transformLayer(layer, sf);
                 // transforming until we reached cases where output scale updated due to situation in downstream layer
                 if (sf.needToRestart()) {
+                    infiniteLoopHistory.back() += "#" + layer->name;
                     break;
                 }
             }
+
+            // looking for infinite loop by using algorithm of compute prefix function, complexity O(N)
+            std::map<int, int> prefixFunction;
+            int k = infiniteLoopHistory.size();
+            for (int i = infiniteLoopHistory.size() - 2; i >= 0; i--) {
+                while (k < infiniteLoopHistory.size() && infiniteLoopHistory[k - 1] != infiniteLoopHistory[i]) {
+                    auto iter = prefixFunction.find(k);
+                    k = iter == prefixFunction.end() ? infiniteLoopHistory.size() : iter->second;
+                }
+
+                if (infiniteLoopHistory[k - 1] == infiniteLoopHistory[i]) {
+                    k--;
+                }
+
+                if ((infiniteLoopHistory.size() - i)%2 == 0 && (infiniteLoopHistory.size() - i)/2 == infiniteLoopHistory.size() - k) {
+                    infiniteLoopPattern.clear();
+                    int patternLength = (infiniteLoopHistory.size() - i)/2;
+                    for (int j = 0; j < patternLength; j++) {
+                        infiniteLoopPattern.emplace_back(infiniteLoopHistory[infiniteLoopHistory.size() - patternLength + j]);
+                    }
+                    infiniteLoopHistory.clear();
+                    gnalog() << "infinite loop detected\n";
+                    break;
+                }
+
+                prefixFunction.emplace(i, k);
+            }
+
+            if (infiniteLoopHistory.empty()) {
+                infiniteLoopCount++;
+            } else {
+                if (infiniteLoopCount > 0 &&
+                    (infiniteLoopHistory.size()%infiniteLoopPattern.size() == 0 || sf.allLayersProcessed()) &&
+                    !std::equal(infiniteLoopHistory.begin() + (infiniteLoopHistory.size() - infiniteLoopPattern.size()),
+                        infiniteLoopHistory.end(), infiniteLoopPattern.begin())) {
+                    infiniteLoopCount = 0;
+                    infiniteLoopPattern.clear();
+                    gnalog() << "infinite loop fixed\n";
+                }
+            }
+
+            sf.SetInfiniteLoopCount(infiniteLoopCount);
+        }
+
+        if (infiniteLoopCount > 0) {
+            THROW_GNA_EXCEPTION << "infinite loop";
```

Comment:
is it possible to add some additional information in the exception description: for example, the infinite loop pattern, or some key layers for this pattern?

Reply:
added